### PR TITLE
deploy controller for tiflash

### PR DESCRIPTION
**Pull request by DanielZhangQD**
### What problem does this PR solve? 
Deploy controller for TiFlash, part of #2050
### What is changed and how does it work?
Deploy TiFlash with CR in TidbCluster.
### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)
   - Deploy TiFlash with one disk
   - Deploy TiFlash with two disks

Code changes

 - Has Go code change

Related changes

 - Need to cherry-pick to the release branch

### Does this PR introduce a user-facing change?:

```release-note
Deploy TiFlash with TidbCluster CR
```

### CODE REVIEW

File: pkg/controller/tidbcluster/tidb_cluster_control.go [hunk 1]
```diff
@@ -180,6 +183,19 @@ func (tcc *defaultTidbClusterControl) updateTidbCluster(tc *v1alpha1.TidbCluster
 		return err
 	}
 
+	// works that should do to making the tiflash cluster current state match the desired state:
+	//   - waiting for the pd cluster available(pd cluster is in quorum)
```
Comment (Yisaer) [inline pkg/controller/tidbcluster/tidb_cluster_control.go:187]: waiting for the tidb cluster

File: pkg/controller/tidbcluster/tidb_cluster_controller.go [hunk 2]
```diff
@@ -106,11 +107,14 @@ func NewController(
 	typedControl := controller.NewTypedControl(controller.NewRealGenericControl(genericCli, recorder))
 	pdScaler := mm.NewPDScaler(pdControl, pvcInformer.Lister(), pvcControl)
 	tikvScaler := mm.NewTiKVScaler(pdControl, pvcInformer.Lister(), pvcControl, podInformer.Lister())
+	tiflashScaler := mm.NewTiKVScaler(pdControl, pvcInformer.Lister(), pvcControl, podInformer.Lister())
 	pdFailover := mm.NewPDFailover(cli, pdControl, pdFailoverPeriod, podInformer.Lister(), podControl, pvcInformer.Lister(), pvcControl, pvInformer.Lister(), recorder)
 	tikvFailover := mm.NewTiKVFailover(tikvFailoverPeriod)
+	tiflashFailover := mm.NewTiKVFailover(tiflashFailoverPeriod)
```
Comment (Yisaer) [inline pkg/controller/tidbcluster/tidb_cluster_controller.go:113]: NewTiFlashFailover

File: pkg/controller/tidbcluster/tidb_cluster_controller.go [hunk 3]
```diff
@@ -106,11 +107,14 @@ func NewController(
...
 	tidbFailover := mm.NewTiDBFailover(tidbFailoverPeriod)
 	pdUpgrader := mm.NewPDUpgrader(pdControl, podControl, podInformer.Lister())
 	tikvUpgrader := mm.NewTiKVUpgrader(pdControl, podControl, podInformer.Lister())
+	tiflashUpgrader := mm.NewTiKVUpgrader(pdControl, podControl, podInformer.Lister())
```
Comment (Yisaer) [inline pkg/controller/tidbcluster/tidb_cluster_controller.go:117]: NewTiFlashUpgrader

File: pkg/controller/tidbcluster/tidb_cluster_controller.go [hunk 4]
```diff
@@ -106,11 +107,14 @@ func NewController(
 	typedControl := controller.NewTypedControl(controller.NewRealGenericControl(genericCli, recorder))
 	pdScaler := mm.NewPDScaler(pdControl, pvcInformer.Lister(), pvcControl)
 	tikvScaler := mm.NewTiKVScaler(pdControl, pvcInformer.Lister(), pvcControl, podInformer.Lister())
+	tiflashScaler := mm.NewTiKVScaler(pdControl, pvcInformer.Lister(), pvcControl, podInformer.Lister())
```
Comment (Yisaer) [inline pkg/controller/tidbcluster/tidb_cluster_controller.go:110]: NewTiFlashScaler

File: pkg/manager/member/tiflash_upgrader.go [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+	return &tiflashUpgrader{
+		pdControl:  pdControl,
+		podControl: podControl,
+		podLister:  podLister,
+	}
+}
+
+func (tku *tiflashUpgrader) Upgrade(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+
+	return nil
```
Comment (Yisaer) [inline pkg/manager/member/tiflash_upgrader.go:43]: Add a TODO here

File: pkg/manager/member/tiflash_scaler.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+func (tfs *tiflashScaler) ScaleIn(tc *v1alpha1.TidbCluster, oldSet *apps.StatefulSet, newSet *apps.StatefulSet) error {
+
+	return nil
+}
+
+// SyncAutoScalerAnn reclaims the auto-scaling-out slots if the target pods no longer exist
+func (tfs *tiflashScaler) SyncAutoScalerAnn(tc *v1alpha1.TidbCluster, actual *apps.StatefulSet) error {
+
+	return nil
+}
```
Comment (Yisaer) [inline pkg/manager/member/tiflash_scaler.go:62]: Add a TODO here

File: pkg/manager/member/tiflash_failover.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+// NewTiFlashFailover returns a tiflash Failover
+func NewTiFlashFailover(tiflashFailoverPeriod time.Duration) Failover {
+	return &tiflashFailover{tiflashFailoverPeriod}
+}
+
+func (tff *tiflashFailover) Failover(tc *v1alpha1.TidbCluster) error {
+	return nil
+}
```
Comment (Yisaer) [inline pkg/manager/member/tiflash_failover.go:33]: Add a TODO here

File: pkg/manager/member/tiflash_util.go [hunk 8]
```diff
@@ -0,0 +1,419 @@
...
+
+const (
+	defaultClusterLog = "/data0/logs/flash_cluster_manager.log"
+	defaultProxyLog   = "/data0/logs/proxy.log"
+	defaultErrorLog   = "/data0/logs/error.log"
+	defaultServerLog  = "/data0/logs/server.log"
+)
+
+func buildTiFlashSidecarContainers(tc *v1alpha1.TidbCluster) []corev1.Container {
+	spec := tc.Spec.TiFlash
```
Comment (Yisaer) [inline pkg/manager/member/tiflash_util.go:34]: Add unit test for these methods?

File: pkg/manager/member/tiflash_member_manager.go [hunk 9]
```diff
@@ -0,0 +1,858 @@
+// Copyright 2018 PingCAP, Inc.
```
Comment (weekface) [inline pkg/manager/member/tiflash_member_manager.go:1]:
```suggestion
// Copyright 2020 PingCAP, Inc.
```

File: pkg/manager/member/tiflash_member_manager.go [hunk 10]
```diff
@@ -0,0 +1,858 @@
...
+
+	oldSvc := oldSvcTmp.DeepCopy()
+
+	equal, err := controller.ServiceEqual(newSvc, oldSvc)
+	if err != nil {
+		return err
+	}
+	if !equal {
+		svc := *oldSvc
+		svc.Spec = newSvc.Spec
```
Comment (weekface) [inline pkg/manager/member/tiflash_member_manager.go:151]:
```suggestion
		svc.Spec = newSvc.Spec
		svc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
```
?

File: pkg/controller/tidbcluster/tidb_cluster_control.go (hunk 1 above)
Comment (DanielZhangQD) [reply pkg/controller/tidbcluster/tidb_cluster_control.go:187]: updated

File: pkg/controller/tidbcluster/tidb_cluster_controller.go (hunk 2 above)
Comment (DanielZhangQD) [reply pkg/controller/tidbcluster/tidb_cluster_controller.go:113]: Updated

File: pkg/controller/tidbcluster/tidb_cluster_controller.go (hunk 3 above)
Comment (DanielZhangQD) [reply pkg/controller/tidbcluster/tidb_cluster_controller.go:117]: Updated

File: pkg/controller/tidbcluster/tidb_cluster_controller.go (hunk 4 above)
Comment (DanielZhangQD) [reply pkg/controller/tidbcluster/tidb_cluster_controller.go:110]: Updated

File: pkg/manager/member/tiflash_upgrader.go (hunk 5 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_upgrader.go:43]: updated

File: pkg/manager/member/tiflash_scaler.go (hunk 6 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_scaler.go:62]: updated

File: pkg/manager/member/tiflash_failover.go (hunk 7 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_failover.go:33]: updated

File: pkg/manager/member/tiflash_member_manager.go (hunk 9 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_member_manager.go:1]: updated

File: pkg/manager/member/tiflash_member_manager.go (hunk 10 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_member_manager.go:151]: This is a headless service, no need to preserve the clusterIP.

File: pkg/manager/member/tiflash_util.go (hunk 8 above)
Comment (DanielZhangQD) [reply pkg/manager/member/tiflash_util.go:34]: updated

**Comment by DanielZhangQD** [comment]
@weekface @Yisaer comments addressed, PTAL again.

**Pull request merged**